File: xskeleton-lib/src/main/java/com/mcgrady/xskeleton/http/handler/RxErrorHandler.java
package com.mcgrady.xskeleton.http.handler;

import android.content.Context;

import com.mcgrady.xskeleton.http.interf.ResponseErrorListener;
import com.mcgrady.xskeleton.http.interf.ResponseResultCodeListener;


/**
 * Created by mcgrady on 2019/4/26.
 */
public class RxErrorHandler {
    public final String TAG = this.getClass().getSimpleName();
    private ErrorHandlerFactory mHandlerFactory;

    private RxErrorHandler(Builder builder) {
        this.mHandlerFactory = builder.errorHandlerFactory;
    }

    public static Builder builder() {
        return new Builder();
    }

    public ErrorHandlerFactory getHandlerFactory() {
        return mHandlerFactory;
    }

    public static final class Builder {
        private Context context;
        private ResponseErrorListener mResponseErrorListener;
        private ErrorHandlerFactory errorHandlerFactory;

        private Builder() {
        }

        public Builder with(Context context) {
            this.context = context;
            return this;
        }

        public Builder responseErrorListener(ResponseErrorListener responseErrorListener) {
            this.mResponseErrorListener = responseErrorListener;
            return this;
        }

        public RxErrorHandler build() {
            if (context == null)
                throw new IllegalStateException("Context is required");

            if (mResponseErrorListener == null)
                throw new IllegalStateException("ResponseErrorListener is required");

            this.errorHandlerFactory = new ErrorHandlerFactory(context, mResponseErrorListener);

            return new RxErrorHandler(this);
        }
    }
}

File: xskeleton-lib/src/main/java/com/mcgrady/xskeleton/http/handler/RxErrorHandler.kt
package com.mcgrady.xskeleton.http.handler

import android.content.Context
import com.mcgrady.xskeleton.http.interf.ResponseErrorListener

/**
 * Created by mcgrady on 2019/4/26.
 */
class RxErrorHandler private constructor(builder: Builder) {
    val TAG = this.javaClass.simpleName
    val handlerFactory: ErrorHandlerFactory?

    class Builder {
        private var context: Context? = null
        private var mResponseErrorListener: ResponseErrorListener? = null
        var errorHandlerFactory: ErrorHandlerFactory? = null
        fun with(context: Context?): Builder {
            this.context = context
            return this
        }

        fun responseErrorListener(responseErrorListener: ResponseErrorListener?): Builder {
            mResponseErrorListener = responseErrorListener
            return this
        }

        fun build(): RxErrorHandler {
            checkNotNull(context) { "Context is required" }
            checkNotNull(mResponseErrorListener) { "ResponseErrorListener is required" }
            errorHandlerFactory = ErrorHandlerFactory(context!!, mResponseErrorListener!!)
            return RxErrorHandler(this)
        }
    }

    companion object {
        fun builder(): Builder {
            return Builder()
        }
    }

    init {
        handlerFactory = builder.errorHandlerFactory
    }
}